utils/dwarf.c

```diff
@@ -907,6 +907,25 @@ static void get_source_location(Dwarf_Die *die, struct build_data *bd,
 	if (dfile == NULL)
 		return;
 
+	Dwarf_Die cudie;
+	dwarf_diecu(die, &cudie, NULL, NULL);
+
+	if (dwarf_hasattr(&cudie, DW_AT_comp_dir)) {
+		char *comp_dir = str_attr(&cudie, DW_AT_comp_dir, false);
+
+		char *d = comp_dir;
+		char *f = dfile->name;
+		while (*d && *f && *d == *f) {
```

Comment:
I think we can compare the string once per compilation unit DIE and reuse it for all children for efficiency.
Please give me your opinion.

Reply:
I'm worry about a big program which might have more than one compilation directories.  In that case we may need to find shortest one and use it for all the places.  So my idea is that do not compare strings here and do it when all srcline info is collected - like in `build_dwarf_info()`.

Reply:
And I think we can simple use `strncmp()`?

Reply:
> about a big program which might have more than one compilation directories. In that case we may need to find shortest one and use it for all the places. 

For understanding, what is shortest one in this case?
A program has 2 CU like as follows.
- first complication dir: /home/soft/uftrace/tests
- second complication dir: /bingbong/test/aaa

Reply:
Yes, that's a case what I'm worring about.  So my idea is like below:

1. save file location as is
2. collect all comp_dir info from all CUs
3. select most used one as a base dir

I think, in many cases, comp_dir will be a child directory of a base like "/home/xxx/uftrace", "/home/xxx/uftrace/tests", "/home/xxx/uftrace/libmcount", "/home/some/unrelated/path".  Then we can count "/home/xxx/uftrace" as 3 (it includes children) and select it.  I think we should not use a part of common directory (in this example "/home") and use a full comp_dir.

Reply:
ok, you want to select 1 base dir for a binary to represent all src file name as a relative path of the base dir, right? (if it has common path with base dir)

If so, It'd be good to discuss the policy for situations below.

1. in case that there's no common full comp_dir.
> I think we should not use a part of common directory (in this example "/home") and use a full comp_dir.

There's no base dir in case below with the policy above. 
Do you agree with it for simplicity?
"/home/xxx/uftrace/cmds", "/home/xxx/uftrace/tests", "/home/xxx/uftrace/libmcount", "/home/some/unrelated/path"

2. in case that used count of dir is same.
in the case like below, do you think we have to use the number of functions of each CU for these case?
- first complication 
dir: /home/soft/uftrace/tests
number of child functions: 1
- second complication 
dir: /bingbong/test/aaa
number of child functions: 100


3. comp_dir of another binary
With https://github.com/namhyung/uftrace/pull/1161, source line of linked library to executable can be shown.
Do you think base dir can be selected independently for each binary?

Reply:
Thanks for sharing your thoughts!

1. if there's no common dir, we can show full path as of now.
2. looks like a good idea.
3. each binary will have a separate base dir.

Reply:
To "select the mostly used one as a base dir", I think we need a routine to compare comp_dir with other comp_dir. 
now I have a version using array for it simply, but time complexity of that could makes delay in big program.
in case of node, the number of comp_dir of main executable has 1673. 
I think the number could be bigger in bigger program. 
```
~/soft/node$ readelf -wi node|grep comp_dir|wc
   1673   13268  167457
```
so, do you think that we need to try to use hash map or trie for this?
there are some light opensource hash tables. followings are used on RTOS device with under 1MB ram.
https://github.com/troydhanson/uthash
https://github.com/rxi/map
Trie can be alternatives, but it could have space complexity issue as condition.

Could you give your opinion?
If available, could you also share the max condition(number of comp_dir / length or dir) you assume or watched?

Reply:
Hmm.. I don't think 1673 is too big and I guess they more mostly repeated.  We can sort and count to have an unique set.. did you already try that?  Or else, while we don't have hash map yet, you can use low-level rbtree APIs or take a look at (and help to get merge) #1051.

Reply:
yes. node is an example. I thought you maybe want uftrace be used in SpaceX.
yes. I have primitive version to see just working simply. I share it anyway for communication.  It's using just array to insert and to search now, it has to be replaced with better data structure.
maybe hashmap could be better than rbtree in performance. but if you agree I'll use your rbtree for this PR, and then replace it with hashmap when #1051 is finished. 

about #1051, it seems that it could to be divided to PR for applying hashmap and PR for using hashmap. it was just an opinion.